pkg/tls/certificate.go

```diff
@@ -273,3 +275,86 @@ func (c *Certificates) Set(value string) error {
 func (c *Certificates) Type() string {
 	return "certificates"
 }
+
+// VerifyPeerCertificate verifies the chain certificates and their URI.
+func VerifyPeerCertificate(uri string, cfg *tls.Config, rawCerts [][]byte) error {
+	// TODO: Refactor to avoid useless verifyChain (ex: when insecureskipverify is false)
+	cert, err := verifyChain(cfg.RootCAs, rawCerts)
+	if err != nil {
+		return err
+	}
+
+	if len(uri) > 0 {
+		return verifyServerCertMatchesURI(uri, cert)
+	}
+
+	return nil
+}
+
+// verifyServerCertMatchesURI is used on tls connections dialed to a server
+// to ensure that the certificate it presented has the correct URI.
+func verifyServerCertMatchesURI(uri string, cert *x509.Certificate) error {
+	if cert == nil {
+		return errors.New("peer certificate mismatch: no peer certificate presented")
+	}
+
+	// Our certs will only ever have a single URI for now so only check that
+	if len(cert.URIs) < 1 {
+		return errors.New("peer certificate mismatch: peer certificate invalid")
+	}
+
+	gotURI := cert.URIs[0]
+
+	// Override the hostname since we rely on x509 constraints to limit ability to spoof the trust domain if needed
+	// (i.e. because a root is shared with other PKI or Consul clusters).
+	// This allows for seamless migrations between trust domains.
+
+	expectURI := &url.URL{}
+	id, err := url.Parse(uri)
+	if err != nil {
+		return fmt.Errorf("%q is not a valid URI", uri)
+	}
+	*expectURI = *id
+	expectURI.Host = gotURI.Host
```

Comment:
@jbdoumenjou Maybe I do not know enough about spiffe, but isn't overriding the host (or even limiting the SAN-URIs to just one) limited to consul-connect? The behavior of this function just feels to specialized to be part of the global API.

Reply:
It's the first implementation, the behavior can change in the future if there are more general use cases.

Reply:
Ok, fine with me. Just wanted to rule out an oversight :+1: